Catch GitHub action errors and report them

andersy005 opened the pull request:
Amends the half-baked #83

kmpaul commented on .github/workflows/register_repositories.py:
```diff
@@ -15,8 +15,33 @@
+    if set(info.keys()) == expected_keys:
+        if info['campaign'] not in config:
+            error_messages.append(
+                f"\n  - Unknown campaign: {info['campaign']}. Valid campaigns include {list(config.keys())}"
```
```suggestion
                f"\n  - Unknown campaign {repr(info['campaign'])}. Valid campaigns include {', '.join(repr(l) for l in config).}"
```

kmpaul commented on .github/workflows/register_repositories.py:
```diff
@@ -15,8 +15,33 @@
+
+        if not info['repo']:
+            valid = False
+            error_messages.append('\n  - No specified repository.')
```
```suggestion
            error_messages.append('\n  - No specified repository')
```

kmpaul commented on .github/workflows/register_repositories.py:
```diff
@@ -15,8 +15,33 @@
+    else:
+        valid = False
+        error_messages.append(
+            f'\n  - Found unexpected keys: {list(info.keys())} in the parsed command. Expected keys are {expected_keys}'
```
```suggestion
            f'\n  - Found unexpected keys ({", ".join(repr(k) for k in info)}) in the parsed command. Expected keys are {", ".join(expected_keys)}.'
```

kmpaul commented on .github/workflows/register_repositories.py:
```diff
@@ -28,23 +53,33 @@ def parse_line(line, original_config, repos={'remove': [], 'add': []}):
+                    config[info['campaign']]['repos'].remove(info['repo'])
+                except ValueError:
+                    error_messages.append(
+                        f"\n  - Unable to remove the following repo: {info['repo']} because it doesn't exist in the list of repos: {config[info['campaign']]['repos']} in the {info['campaign']} campaign."
```
```suggestion
                        f"\n  - Unable to remove the repo {repr(info['repo'])} because it doesn't exist in the list of repos ({', '.join(repr(i) for i in config[info['campaign']]['repos'])}) of the {repr(info['campaign'])} campaign."
```

andersy005 commented:
Great. Thank you for the suggestions! Let me test them locally

andersy005 commented:
The changes produce something along these lines: `('NCAR/jupyter-forward', 'NCAR/ncar-jobqueue')`. Is this `(NCAR/jupyter-forward, NCAR/ncar-jobqueue)` what you were going for instead?

kmpaul commented:
I guess I'm fine leaving off the use of `repr`.  That adds the quotations for strings.  Maybe it reads better without the `repr`, and the code is much shorter, too, because you don't need the generators.

andersy005 commented:
I removed the `repr()` and added markdown backticks instead:

- /remove-repo repo:NCAR/xdevbot-testing campaign:core

  - Unable to remove the repo `NCAR/xdevbot-testing` because it doesn't exist in the list of repos `(NCAR/xdev, NCAR/xdevbot, NCAR/xdev-docker-images, NCAR/test)` of the `core` campaign.

- /remove-repo repo:NCAR/jupyterlab-pbs campaign:platform

  - Unable to remove the repo `NCAR/jupyterlab-pbs` because it doesn't exist in the list of repos `(NCAR/jupyter-forward, NCAR/ncar-jobqueue)` of the `platform` campaign.

- /add-repo repo: campaign:test

  - Unknown campaign `test`. Valid campaigns include `analysis, checkpoint, core, eocb, ingestion, platform, publication, search`. 
  - No specified repository

kmpaul commented:
Very nice!

kmpaul commented on .github/workflows/register_repositories.py:
```diff
@@ -15,8 +15,33 @@
+    else:
+        valid = False
+        error_messages.append(
+            f'\n  - Found unexpected keys `({", ".join(info.keys())})` in the parsed command. Expected keys are {", ".join(expected_keys)}.'
```
Can we back tick the expected keys, too?

andersy005 replied:
Done

andersy005 commented:
@kmpaul, the latest changes produce the follow:

#### 1. Errors comming from the provided commands

- /remove-repo repo:NCAR/xdevbot-testing campaign:core
  - Unable to remove the repo `NCAR/xdevbot-testing` because it doesn't exist in the list of repos `NCAR/xdev, NCAR/xdevbot, NCAR/xdev-docker-images, NCAR/test` of the `core` campaign.

- /remove-repo repo:NCAR/jupyterlab-pbs campaign:platform
  - Unable to remove the repo `NCAR/jupyterlab-pbs` because it doesn't exist in the list of repos `NCAR/jupyter-forward, NCAR/ncar-jobqueue` of the `platform` campaign.

- /add-repo repo: campaign:test
  - Unknown campaign `test`. Valid campaigns include `analysis, checkpoint, core, eocb, ingestion, platform, publication, search`. 
  - No specified repository

Let me know if I missed anything.

kmpaul commented:
The only weird thing is that when displaying the lists of repos the entire list is enclosed in \`\`s.  It's more annoying, but maybe we can change the generation of the list strings from:

```python
f"`{', '.join(repos)}`"
```

to 

```python
f"{', '.join(f'`{r}`' for r in repos)}"
```

...Really, though, I think this is getting really nit-picky.  It looks great, and I can't think of any missing features at the moment.  We can always make fixes later if we want.

andersy005 commented:
> ...Really, though, I think this is getting really nit-picky. It looks great, and I can't think of any missing features at the moment. We can always make fixes later if we want.

😀 Sounds good. Happy to amend this and/or fix bugs in the future
